File: a2/wikipedia_popular.py
from pyspark import SparkConf, SparkContext
import sys

inputs = sys.argv[1]
output = sys.argv[2]

conf = SparkConf().setAppName('word count')
sc = SparkContext(conf=conf)
assert sys.version_info >= (3, 5)  # make sure we have Python 3.5+
assert sc.version >= '2.3'  # make sure we have Spark 2.3+

def words_once(line):
   raw_string = line.split()
   date = raw_string[0]
   language = raw_string[1]
   title = raw_string[2]
   times = raw_string[3]

   if language == "en" and title != "Main_Page" and not title.startswith("Special:"):
       yield (date, (int(times), title))
   
def compare_to(x, y):
    return x if x[0] > y[0] else y

def get_key(kv):
    return kv[0]

def tab_separated(kv):
    return "%s\t%s" % (kv[0], kv[1])

text = sc.textFile(inputs)
words = text.flatMap(words_once)
max_count = words.reduceByKey(compare_to)

outdata = max_count.sortBy(get_key).map(tab_separated)
outdata.saveAsTextFile(output)

File: a2/WikipediaPopular.java
// adapted from https://hadoop.apache.org/docs/current/hadoop-mapreduce-client/hadoop-mapreduce-client-core/MapReduceTutorial.html

import java.io.IOException;
import java.util.StringTokenizer;
import java.util.regex.Pattern;

import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.conf.Configured;
import org.apache.hadoop.fs.Path;
import org.apache.hadoop.io.IntWritable;
import org.apache.hadoop.io.LongWritable;
import org.apache.hadoop.io.Text;
import org.apache.hadoop.mapreduce.Job;
import org.apache.hadoop.mapreduce.Mapper;
import org.apache.hadoop.mapreduce.Reducer;
import org.apache.hadoop.mapreduce.lib.input.TextInputFormat;
import org.apache.hadoop.mapreduce.lib.output.TextOutputFormat;
import org.apache.hadoop.util.Tool;
import org.apache.hadoop.util.ToolRunner;
import org.apache.hadoop.mapreduce.lib.reduce.LongSumReducer;

public class WikipediaPopular extends Configured implements Tool {

	public static class TokenizerMapper
	extends Mapper<LongWritable, Text, Text, LongWritable>{

		private final static LongWritable one = new LongWritable(1L);
		private Text word = new Text();

		@Override
		public void map(LongWritable key, Text value, Context context
				) throws IOException, InterruptedException {
					StringTokenizer rawString = new StringTokenizer(value.toString());

					String date = rawString.nextToken();
					String language = rawString.nextToken();
					String title = rawString.nextToken();
					String times = rawString.nextToken();

					if ((language.equals("en")) && !(title.equals("Main_Page")) && !(title.startsWith("Special:"))){
						word.set(date);
						one.set(Long.valueOf(times));
						context.write(word, one);
					}
			
			}	
		}
	

	public static class LongMaxReducer
	extends Reducer<Text, LongWritable, Text, LongWritable> {
		private LongWritable result = new LongWritable();

		@Override
		public void reduce(Text date, Iterable<LongWritable> values,
				Context context
				) throws IOException, InterruptedException {
			long maxCount = 0L;
			for (LongWritable val : values) {
				maxCount = val.get() > maxCount ? val.get() : maxCount;
			}
			result.set(maxCount);
			context.write(date, result);
		}
	}

	public static void main(String[] args) throws Exception {
		int res = ToolRunner.run(new Configuration(), new WikipediaPopular(), args);
		System.exit(res);
	}

	@Override
	public int run(String[] args) throws Exception {
		Configuration conf = this.getConf();
		Job job = Job.getInstance(conf, "Wikipedia Popular");
		job.setJarByClass(WikipediaPopular.class);

		job.setInputFormatClass(TextInputFormat.class);

		job.setMapperClass(TokenizerMapper.class);
		job.setReducerClass(LongMaxReducer.class);

		job.setOutputKeyClass(Text.class);
		job.setOutputValueClass(LongWritable.class);
		job.setOutputFormatClass(TextOutputFormat.class);
		TextInputFormat.addInputPath(job, new Path(args[0]));
		TextOutputFormat.setOutputPath(job, new Path(args[1]));

		return job.waitForCompletion(true) ? 0 : 1;
	}
}
